cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+	}
+	// if the current firewall state retrieved from the DO firewall API does not match,
+	// we'll need to reconcile the delta between the current firewall (DO API) and the
+	// target firewall (locally cached CCM firewall).
+	if !cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		// update api firewall inbound rules with the expected inbound rules.
+		fwSvc := fm.client.Firewalls
+		fr := &godo.FirewallRequest{
+			Name:          currentFirewall.Name,
+			InboundRules:  svcInboundRules,
+			OutboundRules: currentFirewall.OutboundRules,
+			DropletIDs:    currentFirewall.DropletIDs,
+			Tags:          currentFirewall.Tags,
+		}
+		currentFirewall, resp, err := fwSvc.Update(ctx, currentFirewall.ID, fr)
+		if err != nil {
+			if resp != nil && resp.StatusCode == 404 {
+				// firewall does not exist, so we should create a new one.
+				fr = &godo.FirewallRequest{
+					Name:         fmt.Sprintf(firewallWorkerCCMNameFormat, fm.workerFirewallName),
+					InboundRules: svcInboundRules,
+				}
+				currentFirewall, resp, err = fwSvc.Create(ctx, fr)
+				if err != nil {
+					if resp != nil && resp.StatusCode == 404 {
+						return fmt.Errorf("failed to create new firewall due to status not found response code")
+					}
+					return fmt.Errorf("failed to create new firewall")
+				}
+				// if the firewall api create has succeeded, proceed to update the local firewall
+				// cache.
+				fm.updateCache(currentFirewall)
+				return nil
+			}
+			return fmt.Errorf("failed to add new firewall inbound rules: %s", err)
+		}
+		// if the firewall api update has succeeded, proceed to update the local firewall
+		// cache.
+		fm.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func (fm *firewallManagerOp) updateCache(currentFirewall *godo.Firewall) {
+	fm.fwCache = &firewallCache{
+		firewall: &cachedFirewall{
+			state: currentFirewall,
+		},
+	}
+}
+
+func (fc *FirewallController) onServiceChange(ctx context.Context) error {
+	serviceList, err := fc.serviceLister.List(labels.Nothing())
+	if err != nil {
+		return fmt.Errorf("failed to get service state: %s", err)
+	}
+	inboundRules, err := fc.createInboundRules(serviceList)
+	if err != nil {
+		return fmt.Errorf("failed to create inbound rules on service change: %v", err)
+	}
+	if inboundRules != nil {
+		fm := *fc.firewallManager
+		fm.Set(ctx, inboundRules)
+	}
+	return nil
+}
+
+func (fc *FirewallController) createInboundRules(serviceList []*v1.Service) ([]godo.InboundRule, error) {
+	if serviceList == nil {
+		return nil, fmt.Errorf("failed to retrieve services and their inbound rules")
+	}
+	var nodePortInboundRules []godo.InboundRule
+	for _, svc := range serviceList {
+		if svc.Spec.Type == v1.ServiceTypeNodePort {
+			// this is a nodeport service so we should check for existing inbound rules on all ports.
+			for _, servicePort := range svc.Spec.Ports {
+				nodePortInboundRules = append(nodePortInboundRules, godo.InboundRule{
+					Protocol:  "tcp",
```

Comment:
Just realized we also need a second UDP inbound rule for every TCP rule we make.